pkg/endpoint/bpf.go

```diff
@@ -871,7 +666,213 @@ func (e *Endpoint) realizeBPFState(regenContext *regenerationContext) (compilati
 // The endpoint mutex must not be held.
 // Returns a function which is invoked by the caller to revert some of the
 // state changes (currently proxy configuration) made here.
-func (e *Endpoint) runPreCompilationSteps(owner Owner, regenContext *regenerationContext) error {
+func (e *Endpoint) runPreCompilationSteps(owner Owner, regenContext *regenerationContext) (preCompilationError error) {
+	stats := &regenContext.Stats
+	datapathRegenCtxt := regenContext.datapathRegenerationContext
+
+	stats.waitingForLock.Start()
+	err := e.LockAlive()
+	stats.waitingForLock.End(err == nil)
+	if err != nil {
+		return err
+	}
+
+	defer e.Unlock()
+
+	datapathRegenCtxt.prepareForDatapathRegeneration()
+	currentDir := datapathRegenCtxt.currentDir
+	nextDir := datapathRegenCtxt.nextDir
+
+	// In the first ever regeneration of the endpoint, the conntrack table
+	// is cleaned from the new endpoint IPs as it is guaranteed that any
+	// pre-existing connections using that IP are now invalid.
+	if !e.ctCleaned {
+		go func() {
+			ipv4 := !option.Config.IPv4Disabled
+			created := ctmap.Exists(nil, ipv4, true)
+			if e.ConntrackLocal() {
+				created = ctmap.Exists(e, ipv4, true)
+			}
+			if created {
+				e.scrubIPsInConntrackTable()
+			}
+			close(datapathRegenCtxt.ctCleaned)
+		}()
+	} else {
+		close(datapathRegenCtxt.ctCleaned)
+	}
+
+	// If dry mode is enabled, no further changes to BPF maps are performed
+	if option.Config.DryMode {
+
+		// Compute policy for this endpoint.
+		if err = e.regeneratePolicy(owner); err != nil {
+			return fmt.Errorf("Unable to regenerate policy: %s", err)
+		}
+
+		_ = e.updateAndOverrideEndpointOptions(nil)
+
+		// Dry mode needs Network Policy Updates, but the proxy wait group must
+		// not be initialized, as there is no proxy ACKing the changes.
+		if err, _ = e.updateNetworkPolicy(owner, nil); err != nil {
+			return err
+		}
+
+		if err = e.writeHeaderfile(nextDir, owner); err != nil {
+			return fmt.Errorf("Unable to write header file: %s", err)
+		}
+
+		log.WithField(logfields.EndpointID, e.ID).Debug("Skipping bpf updates due to dry mode")
+		return nil
+	}
+
+	if e.PolicyMap == nil {
+		e.PolicyMap, _, err = policymap.OpenMap(e.PolicyMapPathLocked())
+		if err != nil {
+			return err
+		}
+		// Clean up map contents
+		e.getLogger().Debug("flushing old PolicyMap")
+		err = e.PolicyMap.Flush()
+		if err != nil {
+			return err
+		}
+
+		// Also reset the in-memory state of the realized state as the
+		// BPF map content is guaranteed to be empty right now.
+		e.realizedMapState = make(PolicyMapState)
+	}
+
+	if e.bpfConfigMap == nil {
+		e.bpfConfigMap, _, err = bpfconfig.OpenMapWithName(e.BPFConfigMapPath(), e.BPFConfigMapName())
+		if err != nil {
+			e.Unlock()
```

Comment:
SHould not unlock here!

Reply:
Good catch, will be fixed in next upload.